# HW 10: Practicing SQL Queries
Submitted by: Gideon Tay\
My UNI: gt2528\
Contact me at: gideon.tay@columbia.edu

Let's first load in the SQL extension and connect to the database:

In [1]:
# Load in the SQL extenstion, must have installed ipython-sql first
%load_ext sql

# Connect to database
%sql mysql+pymysql://student:guest@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft

## 1. Getting to know the data
### 1a. Show all tables in the database

In [2]:
%sql SHOW TABLES

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
37 rows affected.


[('accused',),
 ('accused_family',),
 ('appeal',),
 ('calendarcustom',),
 ('case_person',),
 ('commission',),
 ('complaint',),
 ('confession',),
 ('counterstrategy',),
 ('demonicpact',),
 ('denunciation',),
 ('devilappearance',),
 ('elf_fairyelements',),
 ('imprisonment',),
 ('linkedtrial',),
 ('malice',),
 ('mentionedaswitch',),
 ('movestohla',),
 ('musicalinstrument',),
 ('ordeal',),
 ('othercharges',),
 ('othernamedwitch',),
 ('person',),
 ('prevcommission',),
 ('propertydamage',),
 ('ref_parish',),
 ('reference',),
 ('religiousmotif',),
 ('ritualobject',),
 ('shapechanging',),
 ('source',),
 ('torture',),
 ('trial',),
 ('trial_person',),
 ('weathermodification',),
 ('whitemagic',),
 ('witchesmeetingplace',)]

### 1b. Show the first three entries in the table accused.

In [3]:
%%sql 
SELECT * 
FROM accused 
LIMIT 3

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
3 rows affected.


[('A/EGD/10', 'EGD', 10, 'Mareon', 'Quheitt', 'Marion', 'White', None, None, None, 'Female', None, 0, 0, 'Sammuelston', 'P/JO/3539', 'Haddington', 'Haddington', None, None, None, None, None, None, None, None, None, 'SMD', '2001-05-15T11:06:51', 'jhm', '2002-08-09T11:40:51'),
 ('A/EGD/100', 'EGD', 100, 'Thom', 'Cockburn', 'Thomas', 'Cockburn', None, None, None, 'Male', None, 0, 0, None, None, None, 'Haddington', None, None, None, None, None, None, None, None, None, 'SMD', '2001-05-15T11:06:51', 'jhm', '2002-10-02T10:32:30'),
 ('A/EGD/1000', 'EGD', 1000, 'Christian', 'Aitkenhead', 'Christine', 'Aikenhead', None, None, None, 'Female', None, 0, 0, 'Rottinraw', None, None, 'Dumfries', None, None, None, None, None, 'Married', None, None, None, 'SMD', '2001-05-15T11:06:51', 'jhm', '2002-10-01T10:48:12')]

### 1c. How many people are included in the accused table?

In [4]:
%%sql
SELECT COUNT(*)
FROM accused

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(3219,)]

### 1d. Count the number of non-missing values for age in the data.

In [5]:
%%sql
SELECT COUNT(age)
FROM accused

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(166,)]

### 1e. Show a list of unique occupations

In [6]:
%%sql
SELECT DISTINCT occupation
FROM accused

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
32 rows affected.


[(None,),
 ('Servant',),
 ('Vagabond',),
 ('Weaver',),
 ('Midwife',),
 ('Tailor',),
 ('Messenger',),
 ('Brewster',),
 ('Smith',),
 ('Minister',),
 ('Farmer',),
 ('Mealmaker',),
 ('Tasker',),
 ('Maltman',),
 ('Miller',),
 ('Healer',),
 ('Loadman',),
 ('Blacksmith',),
 ('Shop-keeper',),
 ('Stabler',),
 ('Nurse',),
 ('Collier',),
 ('School teacher',),
 ('Slaterer',),
 ('Workman',),
 ('School Master',),
 ('Merchant',),
 ('Fisherman',),
 ('Henwife',),
 ('Cook',),
 ('Sailor',),
 ('Creelman',)]

### 1f. What proportion of accused are female?

Note that we only consider entries where the `sex` column is filled in our `total`. So, the `prop_female` calculated measures the proportion of the accused with known sex who are female. 

We do this instead of defining `total` as the total number of entries, as it could be possible that some of the people with no known sex may be female. We thus exclude them from this calculation.

In [7]:
# Ignore capitalization using 'LOWER()'
n_female = %sql SELECT COUNT(*) FROM accused WHERE LOWER(sex)='female'
total = %sql SELECT COUNT(sex) FROM accused

# Calculate proportion of accused that are female
prop_female = round(n_female[0][0] / total[0][0], 2)

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.
 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


In [8]:
print(f"Proportion of accused that are female is: {prop_female}")

Proportion of accused that are female is: 0.85


## 2. Seeing the Devil

### 2a. List the unique `devil_type`s in the data.

In [9]:
%%sql
SELECT DISTINCT devil_type
FROM devilappearance

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
16 rows affected.


[('Male',),
 ('Female Fairy',),
 ('Male Fairy',),
 ('Animal Devil',),
 ('Spirit',),
 ('Ghost',),
 ('Other Demon',),
 ('Female',),
 (None,),
 ('Fairy',),
 ('Unspecified Devil',),
 ('Insect Devil',),
 ('Baby',),
 ('Human Devil',),
 ('Child Devil',),
 ('Inanimate Object Devil',)]

### 2b. How many of the devil sightings mention the word "black" in the description in the `devil_text` column?

First, save the `devil_text` column as a variable and save the total number of rows as `n_rows`:

In [10]:
# Save devil_text column as a variable
devil_text_col = %sql SELECT devil_text FROM devilappearance

# Save number of rows in the table as n_rows
n_rows = %sql SELECT COUNT(*) FROM devilappearance
n_rows = n_rows[0][0]

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
396 rows affected.
 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


Check through each cell in the `devil_text` column to find the word 'black':

In [11]:
# import regex
import re 

# Initialize a count variable
black_count = 0

# Loop through each cell in devil_text and regex search
for i in range(n_rows):
    if devil_text_col[i][0] is not None: # Only search if not None
        x = re.search(
            r"\bblack\b", # regex pattern to check for
            devil_text_col[i][0], # cell to search
            re.IGNORECASE # ignore capitalization issues
        ) 
        if x:
            black_count += 1

print(f"No. of devil sightings with 'black': {black_count}")

No. of devil sightings with 'black': 120


### 2c. What proportion of the devils (in `devil_type`) are male?

Similar to (1f), we only consider those with non-empty cells in `devil_type` for our total devil count `n_devils`. We then calculate the number of male devils, `n_male`, ignoring capitalization:

In [12]:
# Number of devils with known types (ignore empty cells)
n_devils = %sql SELECT COUNT(devil_type) FROM devilappearance
n_devils = n_devils[0][0]

# Number of male devil_type, ignoring capitalization
n_male = %sql SELECT COUNT(*) FROM devilappearance WHERE LOWER(devil_type)='male'
n_male = n_male[0][0]

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.
 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


Calculate and print the proportion of devils that are male:

In [13]:
prop_male_devil = round(n_male/ n_devils, 2)
print(f"Proportion of devils that are male: {prop_male_devil}")

Proportion of devils that are male: 0.63


## 3. The trial

### 3a. What are the average and maximum numbers of male and female accusers?

In [14]:
%%sql 
SELECT MAX(male_accusers), MAX(female_accusers), 
       AVG(male_accusers), AVG(female_accusers)
FROM trial

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(48.0, 27.0, 0.4181988158304768, 0.2559226932668329)]

We notice that average male and female accuser numbers are less than 1, suggesting 0 accusers for most cases. This does not make a lot of sense. Perhaps our dataset denotes 0 for both `male_accusers` and `female_accusers` if we do not have any data on the accusers.

Let us calculate the average number of accusers again, but this time we ignore cases where both male and femal accuser columns equal zero. In other words, let us only calculate the average number of male and female accusers in cases where the total number of accusers is non-zero:

In [15]:
%%sql
SELECT AVG(male_accusers), AVG(female_accusers)
FROM (
    SELECT male_accusers, female_accusers
    FROM trial
    WHERE male_accusers > 0 OR female_accusers > 0
) AS filtered_set

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(8.182926829268293, 5.0060975609756095)]

### 3b. Count the number of `sentence`s by sentence type. List them in a table (in descending order), excluding missing values. Rename the column headings to something sensible.

In [16]:
%%sql 
SELECT sentence, COUNT(*) AS "No. of cases" # Rename column
FROM trial
WHERE sentence IS NOT NULL # Exclude missing values
GROUP BY sentence
ORDER BY COUNT(*) DESC # List in descending order

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
10 rows affected.


[('Execution', 205),
 ('Released', 52),
 ('Banishment', 27),
 ('Declared Fugitive', 11),
 ('Excommunicated', 6),
 ('Put to the horn', 2),
 ('Hang', 1),
 ('Branded', 1),
 ('Prison', 1),
 ('Public Humiliation', 1)]

### 3c. Do the number of accusers matter for the `verdict`? Compare the average number of accusers by the type of verdict. Again make sure the table is sorted and the headings make sense.



In [17]:
%%sql 
SELECT verdict, 
       AVG(female_accusers + male_accusers) AS "avg no. of accusers"
FROM trial
WHERE verdict IS NOT NULL # Ignore missing verdict
GROUP BY verdict
ORDER BY AVG(female_accusers + male_accusers) DESC

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
4 rows affected.


[('Guilty', 5.804255319148936),
 ('Not Guilty', 4.558139534883721),
 ('Not Proven', 1.0),
 ('Half Guilty', 0.2857142857142857)]

The average number of accusers is the highest for guilty verdicts. Interestingly, half guilty has a lower average number of accusers than not guilty.

Let us do this again, for cases where there is at least one accuser. It doesn't make sense for a case/ trial to not have a single accuser, so maybe we should ignore those cases as having faulty/ incomplete data. If we did so: 

In [18]:
%%sql 
SELECT verdict, 
       AVG(female_accusers + male_accusers) AS "avg no. of accusers"
FROM trial
WHERE verdict IS NOT NULL # Ignore missing verdict
AND (male_accusers > 0 OR female_accusers > 0)
GROUP BY verdict
ORDER BY AVG(female_accusers + male_accusers) DESC

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
4 rows affected.


[('Guilty', 17.94736842105263),
 ('Not Guilty', 10.88888888888889),
 ('Not Proven', 5.5),
 ('Half Guilty', 2.0)]

After ignoring cases with 0 male and female accusers, guilty verdicts still has the highest average number of accusers, followed by not guilty, not proven, then half guilty verdicts.

## 4. Tortured Truth

### 4a. Left join the `trial` and `confession` tables. For what share of trials does the database record confessions? Create a results table with the number of all trials, the number of confessions, and the share of trials with confessions recorded.

**Number of all trials**: Let's start by finding the total number of trials. We find the number of rows in the `trial` table. We also count the number of unique trial reference numbers (`trialref`) in the `trial` table. Since they are equal, this suggests that each row in the table has a distinct value in the `trialref` column, which we can use to uniquely identify each trial.

We have 3211 trials:

In [19]:
%%sql
SELECT COUNT(*) AS n_rows_in_trial_table, 
       COUNT(DISTINCT trialref) AS n_distinct_trialref
FROM trial

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(3211, 3211)]

**Number of confessions**: Next, let's find the total number of confessions in our `confession` table. We can simply count the number of rows in the table. To protect ourselves against the possibility of repeated information across rows, let's also count the number of unique confession reference numbers (`confessionref`) in the table. Since they are equal, this suggests that each row in the table has a distinct value in the `confessionref` column. 

We have 941 confessions:

In [20]:
%%sql 
SELECT COUNT(*) AS n_rows_in_confession_table,
       COUNT(DISTINCT confessionref) AS n_distinct_confessionref
FROM confession

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(941, 941)]

**Find common columns across tables**: If we want to left join the `trial` and `confession` tables, we need to identify common columns that we will use as a basis for the join. To do this, we use the database system catalog's table `information_schema.columns`, which includes metadata on a table, including a column with all of a table's column names (`column_name`).

Below, we find that there are 5 common columns across the `trial` and `confession` table:

In [21]:
%%sql
SELECT column_name AS common_columns
FROM information_schema.columns # contains metadata on columns
WHERE table_name = 'confession'
INTERSECT
SELECT column_name AS common_columns
FROM information_schema.columns
WHERE table_name = 'trial'

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
5 rows affected.


[('createdate',),
 ('createdby',),
 ('lastupdatedby',),
 ('lastupdatedon',),
 ('trialref',)]

We shall use `trialref` as the basis of the join since we have previously shown that it can be used to uniquely identify all the rows (trials) in the `trial` table.

**Share of trials with confessions**: To get the share of trials with confessions, we must know the number of trials with confessions. Note this is different from simply the total number of trials, as each trial may have multiple confessions. 

We can count the number of trials with confessions by conducting a left join of `trial` and `confession` tables on `trialref`, and counting the number of distinct trials (`trialref`) where there was a successful join in that row, meaning that the `confessionref` column is filled.

Then, we can divide the number of trials with confessions by the total number of trials (found to be 3211 earlier) to get the share of trials with confessions:

In [22]:
%%sql
SELECT COUNT(DISTINCT t.trialref) AS n_trials_with_confessions,
       COUNT(DISTINCT t.trialref) * 1.0 / 3211 AS share_with_confessions
FROM trial AS t
LEFT JOIN confession AS c
ON t.trialref = c.trialref # join on the basis of trialref
WHERE c.confessionref IS NOT NULL

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(756, Decimal('0.23544'))]

**Results table**: Let us consolidate all our results of the number of all trials, the number of all confessions, the number of trials with confessions, and the share of trials with confessions in a table.

In [23]:
%%sql
SELECT 
    (SELECT COUNT(DISTINCT trialref) FROM trial) AS n_total_trials,
    (SELECT COUNT(DISTINCT confessionref) FROM confession) AS n_confessions,
    (SELECT COUNT(DISTINCT t.trialref) 
     FROM trial AS t
     LEFT JOIN confession AS c
     ON t.trialref = c.trialref
     WHERE c.confessionref IS NOT NULL) AS n_trials_with_confessions,
    (SELECT COUNT(DISTINCT t.trialref) * 1.0 / n_total_trials
     FROM trial AS t
     LEFT JOIN confession AS c
     ON t.trialref = c.trialref
     WHERE c.confessionref IS NOT NULL) AS share_with_confessions

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(3211, 941, 756, Decimal('0.23544'))]

### 4b. Only a small number of trials have records of torture. Is there a higher share of confessions among trials with records of torture than trials without such record? Hint: You will need to merge on the confession table.

Let's first get an idea of the table `torture`'s contents:

In [24]:
%sql DESCRIBE torture

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
12 rows affected.


[('tortureref', 'text', 'YES', '', None, ''),
 ('torturesystemid', 'text', 'YES', '', None, ''),
 ('tortureid', 'bigint', 'YES', '', None, ''),
 ('trialref', 'text', 'YES', '', None, ''),
 ('tortureexists', 'tinyint(1)', 'YES', '', None, ''),
 ('torturedate', 'text', 'YES', '', None, ''),
 ('torturedate_as_date', 'text', 'YES', '', None, ''),
 ('torturetype', 'text', 'YES', '', None, ''),
 ('createdby', 'text', 'YES', '', None, ''),
 ('createdate', 'text', 'YES', '', None, ''),
 ('lastupdatedby', 'text', 'YES', '', None, ''),
 ('lastupdatedon', 'text', 'YES', '', None, '')]

We immediatedly notice that there is a `trialref` column, which we can use as a basis of joining with the `confession` or `trial` tables. Let's also check if `tortureref` can be used to uniquely identify each torture entry: 

In [25]:
%sql SELECT COUNT(*), COUNT(DISTINCT tortureref) FROM torture

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(110, 110)]

It appears that `tortureref` could be used to identify each unique torture entry, and there are no repeats of torture entries in the `torture` table.

**Share of confessions among trials with torture**: create 2 new tables of unique trials (`trialref`) found in the `torture` and `confession` tables. Then, left join the table of `trials_with_torture` with the `trials_with_confessions` on the basis of `trialref`. Since this is a left join, all trials with torture are kept in the result, so a simple count can give us the total number of trials where there were records of torture.

Counting the number of trials that were sucessfully left joined would then give us the total number of trials where there was both torture and confessions.

Finally, taking their ratio would give us the share of confessions among trials with records of torture:

In [26]:
%%sql
WITH 
    trials_with_torture AS (
        SELECT DISTINCT trialref
        FROM torture
    ),
    trials_with_confessions AS (
        SELECT DISTINCT trialref
        FROM confession
    )
SELECT
    COUNT(DISTINCT tt.trialref) AS n_torture_trials,
    COUNT(DISTINCT tc.trialref) AS n_torture_confession_trials,
    # Multiply by 1.0 to convert it into a float
    COUNT(DISTINCT tc.trialref) * 1.0 / COUNT(DISTINCT tt.trialref) AS share_confessions_with_torture
FROM trials_with_torture AS tt
LEFT JOIN trials_with_confessions AS tc
ON tt.trialref = tc.trialref

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(52, 22, Decimal('0.42308'))]

**Share of confessions among trials without torture**: this time, left join the `trial` table with with the `trials_with_confessions` table on `trialref`, but filter out trials where there is torture (`trialref` found in the `torture` table). The resulting table includes all trials without records of torture. 

Thus, a simple count gives as the number of trials without torture (`n_no_torture_trials`). Since we left joined the trials with confession, counting the number of trials that were sucessfully left joined gives us the total number of trials where there were confessions but no torture. 

Finally, taking the ratio gives us the share of confessions among trials without torture:

In [27]:
%%sql
WITH
    trials_with_confessions AS (
        SELECT DISTINCT trialref
        FROM confession
    )
SELECT 
    COUNT(DISTINCT t.trialref) AS n_no_torture_trials,
    COUNT(DISTINCT ntc.trialref) AS n_no_torture_confession_trials,
    COUNT(DISTINCT ntc.trialref) * 1.0 / COUNT(DISTINCT t.trialref) AS share_confessions_without_torture
FROM trial AS t
LEFT JOIN trials_with_confessions AS ntc
ON t.trialref = ntc.trialref
WHERE t.trialref NOT IN (SELECT trialref FROM torture)

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(3159, 734, Decimal('0.23235'))]

Comparing the results, the share of confessions among trials with records of torture is higher than that among trials without records of torture (0.42308 > 0.23235).

This makes intuitive sense: torture is likely to coerce people into making confessions.